pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -0,0 +1,714 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+	"github.com/openkruise/kruise/pkg/util/fieldindex"
+	"github.com/openkruise/kruise/pkg/util/ratelimiter"
+	"github.com/openkruise/kruise/pkg/util/requeueduration"
+	wsutil "github.com/openkruise/kruise/pkg/util/workloadspread"
+)
+
+func init() {
+	flag.IntVar(&concurrentReconciles, "workloadspread-workers", concurrentReconciles, "Max concurrent workers for WorkloadSpread controller.")
+}
+
+var (
+	concurrentReconciles = 3
+)
+
+const (
+	controllerName = "workloadspread-controller"
+
+	// CreatPodTimeout sets maximum time from the moment a pod is added to CreatePods in WorkloadSpread.Status by webhook
+	// to the time when the pod is expected to be seen by controller. If the pod has not been found by controller
+	// during that time it is assumed, which means it won't be created at all and corresponding record in map can be
+	// removed from WorkloadSpread.Status. It is assumed that pod/ws apiserver to controller latency is relatively small (like 1-2sec)
+	// so the below value should be more enough.
+	CreatPodTimeout = 30 * time.Second
+
+	// DeletePodTimeout is similar to the CreatePodTimeout and it's the time duration for deleting Pod.
+	DeletePodTimeout = 15 * time.Second
+)
+
+var (
+	controllerKruiseKindCS = appsv1alpha1.SchemeGroupVersion.WithKind("CloneSet")
+	controllerKindRS       = appsv1.SchemeGroupVersion.WithKind("ReplicaSet")
+	controllerKindDep      = appsv1.SchemeGroupVersion.WithKind("Deployment")
+	controllerKindJob      = batchv1.SchemeGroupVersion.WithKind("Job")
+)
+
+// this is a short cut for any sub-functions to notify the reconcile how long to wait to requeue
+var durationStore = requeueduration.DurationStore{}
+
+// Add creates a new WorkloadSpread Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles,
+		RateLimiter: ratelimiter.DefaultControllerRateLimiter()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for spec changes to WorkloadSpread
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.WorkloadSpread{}}, &handler.EnqueueRequestForObject{},
+		predicate.GenerationChangedPredicate{})
```

Comment:
`predicate.GenerationChangedPredicate` is not enough, for the controller should reconcile all workloadspread when kruise-manager starts.